#### Create a new notebook, knn_model, and work with the titanic dataset to answer the following:

In [59]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
import sklearn.metrics
import acquire as aq
import prepare
import warnings
warnings.filterwarnings("ignore")

In [3]:
df = aq.get_titanic_data()

In [4]:
df = prepare.prep_titanic(df)
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  \
0         0       3    male  22.0      1      0   7.2500        S   
1         1       1  female  38.0      1      0  71.2833        C   
2         1       3  female  26.0      0      0   7.9250        S   
3         1       1  female  35.0      1      0  53.1000        S   
4         0       3    male  35.0      0      0   8.0500        S   

   embark_town  alone  sex_male  embark_town_Queenstown  \
0  Southampton      0         1                       0   
1    Cherbourg      0         0                       0   
2  Southampton      1         0                       0   
3  Southampton      0         0                       0   
4  Southampton      1         1                       0   

   embark_town_Southampton  
0                        1  
1                        0  
2                        1  
3                        1  
4                        1

In [77]:
df['bl_survival_pred'] = 0

In [6]:
train_val, test = train_test_split(
    df,
    train_size = 0.8,
    random_state = 1349,
    stratify = df['survived']
)

train, validate = train_test_split(
    train_val,
    train_size = 0.7,
    random_state = 1349,
    stratify = train_val['survived']
)

In [7]:
train.shape, validate.shape, test.shape

((498, 14), (214, 14), (179, 14))

In [8]:
x_cols = train.columns.to_list()
y_cols = 'survived'

In [9]:
x_cols.remove('survived')
x_cols.remove('bl_survival_pred')
x_cols.remove('sex')
x_cols.remove('embarked')
x_cols.remove('embark_town')

In [11]:
X_train = train[x_cols]
y_train = train[y_cols]

##### 1. Fit a K-Nearest Neighbors classifier to your training sample and transform (i.e. make predictions on the training sample)

In [10]:
knn = KNeighborsClassifier(n_neighbors=5, weights='uniform')

In [12]:
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [66]:
y_pred = knn.predict(X_train)

##### 2. Evaluate your results using the model score, confusion matrix, and classification report.

In [17]:
print(sklearn.metrics.accuracy_score(y_train, y_pred))

0.8152610441767069


In [21]:
cm = confusion_matrix(y_train, y_pred)
cm

array([[264,  43],
       [ 49, 142]])

In [68]:
print(f'The TP rate is {TP/(TP + FP + TN + FN)}')
print(f'The FP rate is {FP/(TP + FP + TN + FN)}') 
print(f'The TN rate is {TN/(TP + FP + TN + FN)}')
print(f'The FN rate is {FN/(TP + FP + TN + FN)}')

The TP rate is 0.5301204819277109
The FP rate is 0.09839357429718876
The TN rate is 0.285140562248996
The FN rate is 0.08634538152610442


In [22]:
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.86      0.85       307
           1       0.77      0.74      0.76       191

    accuracy                           0.82       498
   macro avg       0.81      0.80      0.80       498
weighted avg       0.81      0.82      0.81       498



##### 3. Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [27]:
mod_1_accuracy = sklearn.metrics.accuracy_score(y_train, y_pred)

In [28]:
print(f'The accuracy of KNN model 1 is {mod_1_accuracy}')

The accuracy of KNN model 1 is 0.8152610441767069


In [69]:
TP = cm[0][0]
FN = cm[0][1]
FP = cm[1][0]
TN = cm[1][1]
print(f'The TP rate is {TP/(TP + FP + TN + FN)}')
print(f'The FP rate is {FP/(TP + FP + TN + FN)}') 
print(f'The TN rate is {TN/(TP + FP + TN + FN)}')
print(f'The FN rate is {FN/(TP + FP + TN + FN)}')

The TP rate is 0.5301204819277109
The FP rate is 0.09839357429718876
The TN rate is 0.285140562248996
The FN rate is 0.08634538152610442


In [31]:
p, r, f1, s = sklearn.metrics.precision_recall_fscore_support(y_train, y_pred, average='binary')

In [32]:
print(f'The precision score is {p}')
print(f'The recall score is {r}') 
print(f'The f1-score is {f1}')
print(f'The support is {s}')

The precision score is 0.7675675675675676
The recall score is 0.743455497382199
The f1-score is 0.7553191489361701
The support is None


##### 4. Run through steps 1-3 setting k to 10

In [33]:
knn2 = KNeighborsClassifier(n_neighbors=10, weights='uniform')

In [34]:
knn2.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=10)

In [61]:
y_pred2 = knn2.predict(X_train)

In [36]:
print(sklearn.metrics.accuracy_score(y_train, y_pred2))

0.7670682730923695


In [37]:
cm2 = confusion_matrix(y_train, y_pred2)
cm2

array([[276,  31],
       [ 85, 106]])

In [38]:
print(classification_report(y_train, y_pred2))

              precision    recall  f1-score   support

           0       0.76      0.90      0.83       307
           1       0.77      0.55      0.65       191

    accuracy                           0.77       498
   macro avg       0.77      0.73      0.74       498
weighted avg       0.77      0.77      0.76       498



In [39]:
mod_2_accuracy = sklearn.metrics.accuracy_score(y_train, y_pred2)

In [40]:
print(f'The accuracy of KNN model 2 is {mod_2_accuracy}')

The accuracy of KNN model 2 is 0.7670682730923695


In [71]:
TP2 = cm2[0][0]
FN2 = cm2[0][1]
FP2 = cm2[1][0]
TN2 = cm2[1][1]
print(f'The TP rate of knn model 2 is {TP2/(TP2 + FP2 + TN2 + FN2)}')
print(f'The FP rate of knn model 2 is {FP2/(TP2 + FP2 + TN2 + FN2)}') 
print(f'The TN rate of knn model 2 is {TN2/(TP2 + FP2 + TN2 + FN2)}')
print(f'The FN rate of knn model 2 is {FN2/(TP2 + FP2 + TN2 + FN2)}')

The TP rate of knn model 2 is 0.5542168674698795
The FP rate of knn model 2 is 0.1706827309236948
The TN rate of knn model 2 is 0.21285140562248997
The FN rate of knn model 2 is 0.06224899598393574


In [42]:
p2, r2, f12, s2 = sklearn.metrics.precision_recall_fscore_support(y_train, y_pred2, average='binary')

In [56]:
print(f'The precision score of knn model 2 is {p2}')
print(f'The recall score of knn model 2 is {r2}') 
print(f'The f1-score of knn model 2 is {f12}')
print(f'The support of knn model 2 is {s2}')

The precision score of knn model 2 is 0.7737226277372263
The recall score of knn model 2 is 0.5549738219895288
The f1-score of knn model 2 is 0.6463414634146342
The support of knn model 2 is None


##### 5. Run through steps 1-3 setting k to 20

In [45]:
knn3 = KNeighborsClassifier(n_neighbors=20, weights='uniform')

In [46]:
knn3.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=20)

In [62]:
y_pred3 = knn3.predict(X_train)

In [48]:
print(sklearn.metrics.accuracy_score(y_train, y_pred3))

0.7409638554216867


In [49]:
cm3 = confusion_matrix(y_train, y_pred3)
cm3

array([[267,  40],
       [ 89, 102]])

In [50]:
print(classification_report(y_train, y_pred3))

              precision    recall  f1-score   support

           0       0.75      0.87      0.81       307
           1       0.72      0.53      0.61       191

    accuracy                           0.74       498
   macro avg       0.73      0.70      0.71       498
weighted avg       0.74      0.74      0.73       498



In [51]:
mod_3_accuracy = sklearn.metrics.accuracy_score(y_train, y_pred3)

In [52]:
print(f'The accuracy of KNN model 3 is {mod_3_accuracy}')

The accuracy of KNN model 3 is 0.7409638554216867


In [70]:
TP3 = cm3[0][0]
FN3 = cm3[0][1]
FP3 = cm3[1][0]
TN3 = cm3[1][1]
print(f'The TP rate of knn model 3 is {TP3/(TP3 + FP3 + TN3 + FN3)}')
print(f'The FP rate of knn model 3 is {FP3/(TP3 + FP3 + TN3 + FN3)}') 
print(f'The TN rate of knn model 3 is {TN3/(TP3 + FP3 + TN3 + FN3)}')
print(f'The FN rate of knn model 3 is {FN3/(TP3 + FP3 + TN3 + FN3)}')

The TP rate of knn model 3 is 0.536144578313253
The FP rate of knn model 3 is 0.178714859437751
The TN rate of knn model 3 is 0.20481927710843373
The FN rate of knn model 3 is 0.08032128514056225


In [54]:
p3, r3, f13, s3 = sklearn.metrics.precision_recall_fscore_support(y_train, y_pred3, average='binary')

In [55]:
print(f'The precision score of knn model 3 is {p3}')
print(f'The recall score of knn model 3 is {r3}') 
print(f'The f1-score of knn model 3 is {f13}')
print(f'The support of knn model 3 is {s3}')

The precision score of knn model 3 is 0.7183098591549296
The recall score of knn model 3 is 0.5340314136125655
The f1-score of knn model 3 is 0.6126126126126126
The support of knn model 3 is None


##### 6. What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why

The precision score of knn model 1 is 0.7675675675675676
The recall score of knn model 1 is 0.743455497382199
The f1-score of knn model 1 is 0.7553191489361701

The precision score of knn model 2 is 0.7737226277372263
The recall score of knn model 2 is 0.5549738219895288
The f1-score of knn model 2 is 0.6463414634146342

The precision score of knn model 3 is 0.7183098591549296
The recall score of knn model 3 is 0.5340314136125655
The f1-score of knn model 3 is 0.6126126126126126

In [73]:
print(sklearn.metrics.accuracy_score(y_train, y_pred))
print(sklearn.metrics.accuracy_score(y_train, y_pred2))
print(sklearn.metrics.accuracy_score(y_train, y_pred3))

0.8152610441767069
0.7670682730923695
0.7409638554216867


- Model 1 performs best on in sample data because thats what the algorithm is built on and it has the least amount of neighbors.

##### 7. Which model performs best on our out-of-sample data from validate?

In [57]:
X_val = validate[x_cols]
y_val = validate[y_cols]

In [74]:
print('Accuracy of Model 1 on validate set: {:.2f}'
     .format(knn.score(X_val, y_val)))
print('Accuracy of Model 2 on validate set: {:.2f}'
     .format(knn2.score(X_val, y_val)))
print('Accuracy of Model 3 on validate set: {:.2f}'
     .format(knn3.score(X_val, y_val)))

Accuracy of Model 1 on validate set: 0.63
Accuracy of Model 2 on validate set: 0.66
Accuracy of Model 3 on validate set: 0.65
